File: src/compiler/regexValidate.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { validateRegex } from "./regexValidate.ts";

test("valid patterns return undefined", () => {
  for (const re of ["a.c", "[a-z]+", "(foo|bar)*", "\\d{3}", "a\\(b", "[a-z&&[^bc]]", "]"]) {
    expect(validateRegex(re)).toBeUndefined();
  }
});

test("unbalanced groups and classes are reported", () => {
  expect(validateRegex("(foo")).toMatch(/unclosed group/);
  expect(validateRegex("foo)")).toMatch(/unmatched/);
  expect(validateRegex("[abc")).toMatch(/unclosed character class/);
});

test("a trailing backslash is reported", () => {
  expect(validateRegex("abc\\")).toMatch(/trailing backslash/);
});

test("parentheses inside a class are literal, not groups", () => {
  expect(validateRegex("[()]")).toBeUndefined();
});

test("\\Q...\\E literal regions are skipped", () => {
  expect(validateRegex("\\Q(unbalanced\\E")).toBeUndefined();
});

File: togo/internal/compiler/regexvalidate_test.go
// Port of src/compiler/regexValidate.test.ts
package compiler

import (
	"strings"
	"testing"
)

func TestRegexValid(t *testing.T) {
	for _, re := range []string{"a.c", "[a-z]+", "(foo|bar)*", `\d{3}`, `a\(b`, "[a-z&&[^bc]]", "]", "[()]"} {
		if reason, bad := ValidateRegex(re); bad {
			t.Errorf("ValidateRegex(%q) = %q, want ok", re, reason)
		}
	}
}

func TestRegexUnbalanced(t *testing.T) {
	cases := map[string]string{
		"(foo": "unclosed group",
		"foo)": "unmatched",
		"[abc": "unclosed character class",
		`abc\`: "trailing backslash",
	}
	for re, want := range cases {
		reason, bad := ValidateRegex(re)
		if !bad || !strings.Contains(reason, want) {
			t.Errorf("ValidateRegex(%q) = (%q,%v), want contains %q", re, reason, bad, want)
		}
	}
}

func TestRegexQuoteRegion(t *testing.T) {
	if _, bad := ValidateRegex(`\Q(unbalanced\E`); bad {
		t.Error(`\Q...\E region should be skipped`)
	}
}
